### PR TITLE
Customize the codecs used in services through sbt

### LINKED ISSUE #371: New Sbt code generation setting
As part of this ticket, we need to add a new sbt setting to be able to specify custom imports including custom marshallers for serializing/deserializing data.

The @fedefernandez's proposal would be:

(Extensible) ADT (provided by frees-rpc atm):

```scala
abstract class MarshallersImport(val marshallersImport: String) extends Product with Serializable
case object BigDecimalAvroMarshallers extends MarshallersImport("import ....")
case object JavaTimeDateAvroMarshallers extends MarshallersImport("import ....")
case object JodaDateTimeAvroMarshallers extends MarshallersImport("import ....")
```

New Sbt Setting:

```scala
lazy val idlGenCodecImports: SettingKey[List[MarshallersImport]]
```

Default value:

```scala
List(BigDecimalAvroMarshallers, JavaTimeDateAvroMarshallers)
```

This new setting would bring the ability to define the custom user marshallers in the `build.sbt`.

* [ ] Sbt Settings.
* [ ] Docs.

Related to #370

### LINKED ISSUE #372: Custom Marshallers in Code Generation
Depends on https://github.com/frees-io/freestyle-rpc/issues/371

We need to modify the current sbt-idl-gen-plugin to be able to do the code generation based on the new sbt setting added as part of #371.

* [ ] Code generation update.
* [ ] Scripted tests.

### PR DESCRIPTION
**Pull request by fedefernandez**
Fixes #371 
Fixes #372 

Currently, we've got both *generators* (`avro` and `idl`) under the same "`trait`" and because the imports are something specific of the `AvroSrcGenerator` I've decided to pass the new property `idlGenCodecImports` through the constructor instead of through the task params like [the options](https://github.com/frees-io/freestyle-rpc/compare/ff-371-sbt-codec-imports-generation?expand=1#diff-78d1790b16820366c416d210b1eb3015R158). Please, let me know if that makes sense.

### CODE REVIEW

File: modules/idlgen/core/src/main/scala/Model.scala [hunk 1]
```diff
@@ -63,4 +63,16 @@ object Model {
...
+      extends Product
+      with Serializable
+
+  final case class CustomMarshallersImport(mi: String) extends MarshallersImport(mi)
+  case object BigDecimalAvroMarshallers
+      extends MarshallersImport("freestyle.rpc.internal.encoders.avro.bigdecimal._")
+  case object JavaTimeDateAvroMarshallers
+      extends MarshallersImport("freestyle.rpc.internal.encoders.avro.javatime._")
+  case object JodaDateTimeAvroMarshallers
+      extends MarshallersImport("freestyle.rpc.marshallers.jodaTimeEncoders.avro._")
```
Comment (juanpedromoreno) [inline modules/idlgen/core/src/main/scala/Model.scala]: Should we add here the proto imports?

File: modules/idlgen/core/src/main/scala/avro/AvroSrcGenerator.scala [hunk 2]
```diff
@@ -105,11 +109,10 @@ object AvroSrcGenerator extends SrcGenerator {
...
 
-    val importLines = Seq(
-      "import freestyle.rpc.internal.encoders.avro.bigdecimal._",
-      "import freestyle.rpc.internal.encoders.avro.javatime._",
-      "import freestyle.rpc.protocol._"
-    )
+    val importLines =
+      ("import freestyle.rpc.protocol._" :: marshallersImports
+        .map(_.marshallersImport)
+        .map("import " + _)).sorted
```
Comment (juanpedromoreno) [inline modules/idlgen/core/src/main/scala/avro/AvroSrcGenerator.scala]: Should we check if the user has specified the `import` reserved word before adding it?

File: modules/idlgen/plugin/src/main/scala/IdlGenPlugin.scala [hunk 3]
```diff
@@ -108,7 +114,8 @@ object IdlGenPlugin extends AutoPlugin {
     srcGenSourceDirs := Seq(srcGenSourceDir.value),
     srcGenIDLTargetDir := (Compile / resourceManaged).value / idlType.value,
     srcGenTargetDir := (Compile / sourceManaged).value,
-    genOptions := Seq.empty
+    genOptions := Seq.empty,
+    idlGenCodecImports := List(BigDecimalAvroMarshallers, JavaTimeDateAvroMarshallers)
```
Comment (juanpedromoreno) [inline modules/idlgen/plugin/src/main/scala/IdlGenPlugin.scala]: Are we tying the default values to Avro? Do you think we could the `idlType` setting to figure out these values?

File: modules/idlgen/plugin/src/sbt-test/sbt-frees-rpc-idlgen/importMarshallers/project/plugins.sbt [hunk 4]
```diff
@@ -0,0 +1 @@
+addSbtPlugin("io.frees" %% "sbt-frees-rpc-idlgen" % "0.14.2-SNAPSHOT")
```
Comment (juanpedromoreno) [inline modules/idlgen/plugin/src/sbt-test/sbt-frees-rpc-idlgen/importMarshallers/project/plugins.sbt]: version should be `sys.props("version")`

File: modules/idlgen/core/src/main/scala/Model.scala [hunk 5]
```diff
@@ -63,4 +63,16 @@ object Model {
       case _ => false
     }
   }
+
+  sealed abstract class MarshallersImport(val marshallersImport: String)
```
[the remark this answers is not in the document]
Comment (fedefernandez) [reply modules/idlgen/core/src/main/scala/Model.scala]: There is a custom one for that:
https://github.com/frees-io/freestyle-rpc/pull/374/files#diff-9adb97b8bcc89ae7f61c2758ce7c7583R17

File: modules/idlgen/core/src/main/scala/Model.scala (hunk 1 above)
Comment (fedefernandez) [reply modules/idlgen/core/src/main/scala/Model.scala]: They are not supported, but we should, yeah

File: docs/src/main/tut/idl-generation.md [hunk 6]
```diff
@@ -221,6 +221,8 @@ Just like `idlGen`, `srcGen` and `srcGenFromJars` has some configurable settings
   * All the source folders specified in the `srcGenSourceDirs` setting.
 * **`srcGenTargetDir`**: the Scala target base directory, where the `srcGen` task will write the Scala files in subdirectories/packages based on the namespaces of the IDL files. By default: `Compile / sourceManaged`, typically `target/scala-2.12/src_managed/main/`.
 * **`genOptions`**: additional options to add to the generated `@service` annotations, after the IDL type. Currently only supports `"Gzip"`.
+* **`idlGenCodecImports`**: additional imports to add on top to the generated service files. This property can be used for importing extra encoders/decoders for your services. By default: `List(BigDecimalAvroMarshallers, JavaTimeDateAvroMarshallers)`. The `JodaDateTimeAvroMarshallers` is also available, but it needs the dependency `frees-rpc-marshallers-jodatime`. You can also specify custom imports with the following:
```
Comment (juanpedromoreno) [inline docs/src/main/tut/idl-generation.md]: Maybe we can rename this to `idlGenMarshallerImports`, to be fully aligned, thoughts?

File: modules/idlgen/plugin/src/main/scala/IdlGenPlugin.scala [hunk 7]
```diff
@@ -90,6 +91,11 @@ object IdlGenPlugin extends AutoPlugin {
     lazy val genOptions: SettingKey[Seq[String]] =
       settingKey[Seq[String]](
         "Options for the generator, such as additional @service annotation parameters in srcGen.")
+
+    lazy val idlGenCodecImports: SettingKey[List[MarshallersImport]] =
```
Comment (juanpedromoreno) [inline modules/idlgen/plugin/src/main/scala/IdlGenPlugin.scala]: `idlGenMarshallerImports`?

File: docs/src/main/tut/idl-generation.md (hunk 6 above)
Comment (fedefernandez) [reply docs/src/main/tut/idl-generation.md]: Agree

File: modules/idlgen/core/src/test/scala/AvroScalaGeneratorArbitrary.scala [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+         |
+         |  def sayHelloAvro(arg: foo.bar.HelloRequest): F[foo.bar.HelloResponse]
+         |
+         |  def sayNothingAvro(arg: Empty.type): F[Empty.type]
+         |
+         |}""".stripMargin.split("\n").filter(_.length > 0).toList
+  }
+
+  val importSliceGen: Gen[String] =
+    Gen.choose(4, 10).flatMap(Gen.listOfN(_, Gen.choose('a', 'z')).map(_.mkString("")))
```
[the remark this answers is not in the document]
Comment (fedefernandez) [reply modules/idlgen/core/src/test/scala/AvroScalaGeneratorArbitrary.scala]: I want to keep the generated string bounded by some logic limits, I'll use the `alphaLowerChar` 👍

**Pull request merged**